### DISCUSSION TITLE
itables (datatables.net) in marimo

### DISCUSSION THREAD

**Discussion by liquidcarbon** [marimo-team/marimo#3192] [Ideas]
I've long been a fan of jQuery-based [DataTables.net](https://datatables.net/) and their pythonic wrapper [itables](https://github.com/mwouts/itables).

They're extremely powerful and flexible and are significantly more performant compared to the tables offered in marimo.

I was not able to get them to render them inside Marimo directly, but you they can work in an iframe.

The iframe can also be pointed at another FastAPI endpoint serving such table in a mounted app scenario.
Was not able to make it work via `iframe srcdoc`.

```
mo.Html("""
<iframe src="https://liquidcarbon-duckdb-fastapi-gradio.hf.space/q?query_id=01HWDNFA8QTEA1Y1V229501RMF"
style="width: 900px; min-height: 540px;">
</iframe>
""")
```

![image](https://github.com/user-attachments/assets/7b0bf93e-bd52-4f0e-b233-0f655fdeb4b3)

Just going to leave this here in case others have feelings about these tables.

**Comment by akshayka** [comment]
@liquidcarbon In what way are they more performant? Or rather, in what areas could the performance of marimo's native tables be improved?

**Reply by liquidcarbon** [reply]
If you try to display a dataframe with 10K rows, DataTables handles it easily (purely client side, can do 100K+ with SSR; YMMV depending on number of cols);
Marimo would take a while to respond.

**Comment by mwouts** [comment]
Thank you @liquidcarbon for pointing out at Marimo and for your kind words on ITables!

I see that the [ITable widget](https://mwouts.github.io/itables/widget.html) from ITables v2.2.4 works in Marimo to some extend - give a try to the below:
```
import pandas as pd

from itables.widget import ITable

df = pd.read_json(
    "https://raw.githubusercontent.com/vega/vega-datasets/master/data/cars.json"
)

ITable(df)
```

I have not tested it thoroughly yet but I suspect already that some additional work will be required to have the widget work properly in Marimo. In a quick test I have seen the following problems:
- the table widget is not updated when I modify the underlying dataframe
- the selected rows are not highlighted.

Maybe the Marimo developers will have some documentation I can look at re how to best integrate a widget in this ecosystem?

**Reply by mscolnick** [reply]
@mwouts awesome to see it mostly works

> the table widget is not updated when I modify the underlying dataframe

not sure about this issue. let me know if you need help debugging. 

> the selected rows are not highlighted.

we place widgets in a ShadowDOM so may change the structure you expect when writing your CSS selectors

**Reply by mwouts** [reply]
> the table widget is not updated when I modify the underlying dataframe

That was an issue with the ITable widget, it got fixed in v2.4.0 (https://github.com/mwouts/itables/issues/376)

**Comment by mwouts** [comment]
Hi @mscolnick , I have released a new version of ITables (v2.4.0) which makes the ITable widget as functional as the show method. The ITable widget is now the recommended way to show a table in a Marimo notebook.

However, we do see some problems when rendering tables within Marimo:
- https://github.com/mwouts/itables/issues/387
- https://github.com/mwouts/itables/issues/383

Re the selected row issue I see your comment:

> we place widgets in a ShadowDOM so may change the structure you expect when writing your CSS selectors

but I am unfortunately not very skilled with CSS... do you think you could help us a bit re these two issues ? Thank you !

**Reply by mscolnick** [reply]
hey @mwouts - it looks like esbuild's css import just puts the css on the main document (not handling shadow doms) https://esbuild.github.io/content-types/#css-from-js

i think if you use `@import` in the widget.css [instead of import directly in js](https://github.com/mwouts/itables/blob/main/packages/itables_anywidget/js/widget.ts#L5), that might fix the issue.

this is because if your css goes through the `_css` traitlet, we do put it on the shadowdom for you

**Reply by mwouts** [reply]
Great suggestion, thank you @mscolnick ! I have implemented that in a [PR](https://github.com/mwouts/itables/pull/388), which does provide a meaningfull `._css` traitless, but that does not seem to fix the issue yet - and I don't see that css under the shadow dom object either. Do you see what else I could do? Thank you!
